### PR TITLE
Added option 'Remove perfect albums (allowing multiple file extensions)'

### PR DESCRIPTION
**Pull request by dsparkplug**
This option treats albums with multiple complete sets of track files but with different extensions as perfect.

See #222

### CODE REVIEW

File: plugins/remove_perfect_albums/remove_perfect_albums.py [hunk 1]
```diff
@@ -19,4 +22,65 @@ def callback(self, objs):
               	and album.get_num_unsaved_files() == 0 and album.loaded == True):
                 self.tagger.remove_album(album)
 
+class RemovePerfectAlbumsAllowingMultipleFileExtensions(BaseAction):
+    NAME = 'Remove perfect albums (allowing multiple file extensions)'
+
+    def callback(self, objs):
+        for album in objs:
+            
+            if (not(isinstance(album, Album))):
```
Comment (zas) [inline plugins/remove_perfect_albums/remove_perfect_albums.py:31]: You don't need to add parenthesis here:

```python
if not isinstance(album, Album):
```

File: plugins/remove_perfect_albums/remove_perfect_albums.py [hunk 2]
```diff
@@ -19,4 +22,65 @@ def callback(self, objs):
               	and album.get_num_unsaved_files() == 0 and album.loaded == True):
                 self.tagger.remove_album(album)
 
+class RemovePerfectAlbumsAllowingMultipleFileExtensions(BaseAction):
+    NAME = 'Remove perfect albums (allowing multiple file extensions)'
+
+    def callback(self, objs):
+        for album in objs:
+
```
Comment (zas) [inline plugins/remove_perfect_albums/remove_perfect_albums.py:30]: This whiteline should be removed. See PEP8 for Python style recommendations.

File: plugins/remove_perfect_albums/remove_perfect_albums.py [hunk 3]
```diff
@@ -19,4 +22,65 @@ def callback(self, objs):
...
+    def callback(self, objs):
+        for album in objs:
+            
+            if (not(isinstance(album, Album))):
+                continue
+           
+            title = album.column('title')
+            log.info('Checking album perfection: %s', title)
+            
+            if (album.loaded != True):
```
Comment (zas) [inline plugins/remove_perfect_albums/remove_perfect_albums.py:37]: Useless parenthesis and test can be done this way:

```python
if album.loaded:
```

File: plugins/remove_perfect_albums/remove_perfect_albums.py [hunk 4]
```diff
@@ -19,4 +22,65 @@ def callback(self, objs):
...
+                log.info('Album is not loaded')
+                continue
+
+            unsavedFiles = album.get_num_unsaved_files()
+            if (unsavedFiles > 0):
```
Comment (zas) [inline plugins/remove_perfect_albums/remove_perfect_albums.py:42]:
```python
if unsavedFiles:
```

File: plugins/remove_perfect_albums/remove_perfect_albums.py [hunk 5]
```diff
@@ -19,4 +22,65 @@ def callback(self, objs):
...
+                log.info('Album has %s unsaved files', unsavedFiles)
+                continue
+            
+            unmatchedFiles = album.get_num_unmatched_files()
+            if (unmatchedFiles > 0):
```
Comment (zas) [inline plugins/remove_perfect_albums/remove_perfect_albums.py:47]:
```python
if unmatchedFiles:
```

File: plugins/remove_perfect_albums/remove_perfect_albums.py [hunk 6]
```diff
@@ -19,4 +22,65 @@ def callback(self, objs):
...
+                continue
+
+            if (totalFiles % numberOfTracks != 0):
+                log.info('The number of files %s in album is not wholly divisible by the number of tracks %s', totalFiles, numberOfTracks)
+                continue
+
+            #get dictionary of file extension counts
+            extDict = {}
+            for file in album.iterfiles():
+                _, ext = os.path.splitext(file.filename)
```
Comment (zas) [inline plugins/remove_perfect_albums/remove_perfect_albums.py:64]: Do not use `_` as placeholder, it is used for translations (`_()` is a function), just use `unused`

**Review by zas — CHANGES_REQUESTED** [review]
Please respect [PEP8](https://www.python.org/dev/peps/pep-0008/)

I didn't test the PR yet, but I will soon.

**Comment by zas** [comment]
@dsparkplug any update?

**Comment by Sophist-UK** [comment]
IMO this is trying to do in a plugin what should be done in Picard proper i.e. because we need the icons to reflect the perfect nature of the album or otherwise - and this PR will essentially remove albums which are shown as needing saving in the UI. We need Picard to recognise perfect albums which have multiple formats.

**Comment by phw** [comment]
As this hasn't gone anywhere I'm going to close this for now. Can be re-opened if needed.

And I agree with @Sophist-UK, we could at least have an option to treat a release with multiple files per track as completed. Maybe at least if each track has the same amount of files and they are of different formats.

**Pull request closed without merging**